Use RSSReport class

gita-omr opened the pull request:
Depends on: https://github.com/eclipse/omr/pull/7349

dsouzai commented:
I'm not sure it makes sense for this to go into 0.46 since this seems to be a non functional (i.e., reporting) feature rather than a bug fix, esp since the depending OMR change likely needs some non-trivial changes. Is there a strong reason to get this into 0.46?

gita-omr commented:
> I'm not sure it makes sense for this to go into 0.46 since this seems to be a non functional (i.e., reporting) feature rather than a bug fix, esp since the depending OMR change likely needs some non-trivial changes. Is there a strong reason to get this into 0.46?

I would say it does add a new feature (under an option). We planned to get it into 0.46 - just were a bit late.

pshipton commented:
FYI there are conflicts that need to be resolved before this is merged. 

I don't have any objection to putting this in 0.46 as long as this is merged to master and there are no resulting problems. My understand is it's disabled by default, we haven't started any Milestone builds for 0.46 yet, and we're well in advance of M2. If we are putting it in, it would be nice to have it in before we start M1 builds next week.

gita-omr commented:
Addressed OMR comments and rebased.

gita-omr commented:
Made loop in addRSSItem more sequential to avoid an extra store of the item size into memory.

mpirvu commented on runtime/compiler/control/rossa.cpp:
```diff
@@ -1396,6 +1396,12 @@ onLoadInternal(
+   bool produceRSSReport = produceRSSReportDetailed || TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseRSSReport);
+
+   if (produceRSSReport)
+      new (PERSISTENT_NEW) OMR::RSSReport(produceRSSReportDetailed);
```
The output of this is not stored anywhere. I am guessing the constructor does some special magic, but this is rather unorthodox.

gita-omr replied:
I think it's relatively common for singleton classes...

mpirvu commented on runtime/compiler/build/files/common.mk:
```diff
@@ -392,7 +392,8 @@ JIT_PRODUCT_SOURCE_FILES+=\
     omr/compiler/runtime/OMRCodeCacheMemorySegment.cpp \
-    omr/compiler/runtime/OMRRuntimeAssumptions.cpp
+    omr/compiler/runtime/OMRRuntimeAssumptions.cpp \
+    omr/compiler/runtime/OMRRSSReport.cpp
```
I vaguely recall that there is another make file where you need to list the newly created files to be compiled

gita-omr replied:
Maybe in OMR? I can build OpenJ9.

mpirvu commented on runtime/compiler/runtime/DataCache.cpp:
```diff
@@ -293,6 +293,19 @@ TR_DataCache* TR_DataCacheManager::allocateNewDataCache(uint32_t minimumSize)
                dataCache->_vmThread = NULL;
                dataCache->_allocationMark = dataCacheSeg->heapAlloc;
+
+#ifdef LINUX
```
This code is protected by #ifdef Linux, but in other files in this PR I have not seen the same ifdef protection.

gita-omr replied:
You are right. I better remove it for clarity.

mpirvu commented on runtime/compiler/runtime/DataCache.cpp:
```diff
@@ -293,6 +293,19 @@ TR_DataCache* TR_DataCacheManager::allocateNewDataCache(uint32_t minimumSize)
+                  J9JavaVM * javaVM = jitConfig->javaVM;
+                  PORT_ACCESS_FROM_JAVAVM(javaVM); // for j9vmem_supported_page_sizes
+
+                  dataCache->_rssRegion = new (PERSISTENT_NEW) OMR::RSSRegion("data cache", dataCacheSeg->heapBase, segSize, OMR::RSSRegion::lowToHigh,
```
dataCache->_rssRegion needs to be set to NULL, if it's not used.

mpirvu replied:
What happens if the allocation fails?

gita-omr replied:
Do you mean if allocation of` OMR::RSSRegion` fails?

gita-omr replied:
Added default constructor and initiliazed `_rssRegion`. Thanks for noticing!

mpirvu commented on runtime/compiler/runtime/DataCache.cpp:
```diff
@@ -741,7 +755,7 @@ extern "C" {
 }
 
-//#define DEBUG_DISCLAIM
+#define DEBUG_DISCLAIM  // GITA
```
Please remove "// GITA"

gita-omr replied:
Done

mpirvu commented on runtime/compiler/runtime/J9CodeCache.hpp:
```diff
@@ -36,6 +36,7 @@ namespace J9 { typedef CodeCache CodeCacheConnector; }
 #include "env/IO.hpp"
 #include "env/VMJ9.h"
 
+
```
Please eliminate this new line

gita-omr replied:
Done

mpirvu commented on runtime/compiler/runtime/J9CodeCache.cpp:
```diff
@@ -146,6 +146,19 @@ J9::CodeCache::initialize(TR::CodeCacheManager *manager,
+      J9JavaVM * javaVM = jitConfig->javaVM;
+      PORT_ACCESS_FROM_JAVAVM(javaVM); // for j9vmem_supported_page_sizes
+
+      _coldCodeRSSRegion = new (PERSISTENT_NEW) OMR::RSSRegion("cold cache", _coldCodeAlloc, 0,
```
I am guessing `_coldCodeRSSRegion` is defined in OMR. Is it initialized to NULL?

gita-omr replied:
Yes

dsouzai replied:
It may be worth asserting that `_coldCodeRSSRegion` is null to catch a potential future change in OMR that could allocate some (persistent) memory that then gets leaked by this code.

gita-omr replied:
Done

gita-omr commented:
Thanks for the review. I will address it shortly. I decided to make this PR WIP and include ability to use this class from VM as well.

gita-omr commented on runtime/compiler/runtime/DataCache.hpp:
```diff
@@ -41,6 +42,7 @@ class TR_DataCache
    uint8_t         *_allocationMark; // used if we want to give back memory up to previously set mark
    //TR::Monitor       *_mutex;     // Is this needed?
    int32_t          _status;    // mostly RAS at this point
+   OMR::RSSRegion  *_rssRegion;
```
Done

gita-omr commented:
Rebased.

gita-omr commented:
I think I addressed all the comments and I am not planning on adding any new interfaces to that class right now. @dsouzai @mpirvu  could you please take another look?

dsouzai commented on runtime/compiler/runtime/DataCache.cpp:
```diff
@@ -741,7 +753,7 @@ extern "C" {
 }
 
-//#define DEBUG_DISCLAIM
+#define DEBUG_DISCLAIM
```
Did you mean uncomment this?

gita-omr replied:
Fixed. Thanks!

dsouzai commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -4252,6 +4252,7 @@ typedef struct J9JITConfig {
 	uint64_t clientUID;
 	uint64_t serverUID;
 #endif /* J9VM_OPT_JITSERVER */
+        void (*jitAddNewLowToHighRSSRegion)(const char *name, uint8_t *start, uint32_t size, size_t pageSize);
```
This should use a tab rather than spaces.

gita-omr replied:
Fixed

mpirvu commented on runtime/compiler/runtime/DataCache.hpp:
```diff
@@ -41,13 +42,16 @@ class TR_DataCache
       FULL,
    };
+
+   TR_DataCache() : _rssRegion(NULL) {}
```
The constructor will not be called. It's better to initialize this field at the point where all the others are initialized.

gita-omr replied:
Done

gita-omr commented:
Track the whole Code Cache vs just cold part.

gita-omr commented:
Addressed latest comments

gita-omr commented:
Rebased

mpirvu commented on runtime/compiler/runtime/DataCache.hpp:
```diff
@@ -41,13 +42,17 @@ class TR_DataCache
    };
+
+   // Note: the constructor is not called
+   TR_DataCache() {}
```
What is the purpose of adding this empty constructor?

gita-omr replied:
That was more for the comment. To make it clear that the constructor is not called. But I guess I can remove the constructor and just leave the comment.

gita-omr replied:
Removed the constructor, but left the comment.

mpirvu commented on runtime/compiler/control/HookedByTheJit.cpp:
```diff
@@ -4702,6 +4714,23 @@ void memoryDisclaimLogic(TR::CompilationInfo *compInfo, uint64_t crtElapsedTime,
 #endif // J9VM_INTERP_PROFILING_BYTECODES
    }
 
+void rssReportLogic(TR::CompilationInfo *compInfo)
```
Who is calling this new function?

gita-omr replied:
I added this comment:
`// this method can be used to produce regular RSS reports`

gita-omr commented:
Added check for `rssReport != NULL` inside `jitAddNewLowToHighRSSRegion().`

gita-omr commented:
Addressed latest comments

mpirvu commented:
The commits should be squashed into one.

gita-omr commented:
Squashed all the commits.

dsouzai commented:
jenkins test sanity.functional all jdk21

dsouzai commented:
@gita-omr the builds are failing due to:
```
[2024-07-11T14:05:10.969Z] In file included from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/control/CompilationThread.hpp:38,
[2024-07-11T14:05:10.969Z]                  from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/net/ServerStream.hpp:29,
[2024-07-11T14:05:10.969Z]                  from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/control/CompilationRuntime.hpp:45,
[2024-07-11T14:05:10.969Z]                  from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/../compiler/codegen/J9CodeGenerator.hpp:45,
[2024-07-11T14:05:10.969Z]                  from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/x/codegen/J9CodeGenerator.hpp:26,
[2024-07-11T14:05:10.969Z]                  from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/x/amd64/codegen/J9CodeGenerator.hpp:39,
[2024-07-11T14:05:10.969Z]                  from /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/codegen/CodeGenGC.cpp:34:
[2024-07-11T14:05:10.970Z] /home/jenkins/workspace/Build_JDK21_x86-64_linux_Personal/openj9/runtime/compiler/runtime/DataCache.hpp:29:10: fatal error: runtime/OMRRSSReport.hpp: No such file or directory
[2024-07-11T14:05:10.970Z]    29 | #include "runtime/OMRRSSReport.hpp"
[2024-07-11T14:05:10.970Z]       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
[2024-07-11T14:05:10.970Z] compilation terminated.
```

dsouzai commented:
This PR will conflict with https://github.com/eclipse-openj9/openj9/pull/19850 once that's been merged.

gita-omr commented:
Rebased and removed dependency from the description.

gita-omr commented:
Not sure why `OMRRSSReport.hpp`  was not found. Maybe eclipse/omr#7349 was not picked up by the build yet.

dsouzai commented:
jenkins test sanity.functional all jdk21

gita-omr commented:
The windows build failure seems to be unrelated.
